lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs

```diff
@@ -0,0 +1,907 @@
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( SeqState (..) )
+import Cardano.Wallet.Primitive.Model
+    ( Wallet )
+import Cardano.Wallet.Primitive.Types
+    ( Hash (..), Tx (..), TxMeta (..), WalletId (..), WalletMetadata (..) )
+import Control.Foldl
+    ( Fold (..) )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Crypto.Hash
+    ( hash )
+import Data.Bifunctor
+    ( bimap, first )
+import Data.Foldable
+    ( toList )
+import Data.Functor.Classes
+    ( Eq1, Show1 )
+import Data.Map
+    ( Map )
+import Data.Maybe
+    ( catMaybes, fromJust, fromMaybe, isJust, isNothing )
+import Data.Set
+    ( Set )
+import Data.TreeDiff
+    ( ToExpr (..), defaultExprViaShow )
+import GHC.Generics
+    ( Generic )
+import System.Random
+    ( getStdRandom, randomR )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Args (..)
+    , Gen
+    , Property
+    , collect
+    , elements
+    , frequency
+    , labelledExamplesWith
+    , property
+    , (===)
+    )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+import Test.QuickCheck.Random
+    ( mkQCGen )
+import Test.StateMachine
+    ( CommandNames (..)
+    , Concrete
+    , GenSym
+    , Logic (..)
+    , Reason (..)
+    , Reference
+    , StateMachine
+    , Symbolic
+    , elem
+    , forAllCommands
+    , forAllParallelCommands
+    , forall
+    , prettyCommands
+    , prettyParallelCommands
+    , runCommands
+    , runParallelCommands
+    , (.==)
+    )
+import Test.StateMachine.Types
+    ( Command (..), Commands (..), ParallelCommands, ParallelCommandsF (..) )
+
+import qualified Control.Foldl as Foldl
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.List as L
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Test.QuickCheck as QC
+import qualified Test.StateMachine.Types as QSM
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+
+{-# ANN module ("HLint: ignore Unused LANGUAGE pragma" :: String) #-}
+
+{-------------------------------------------------------------------------------
+  Errors
+-------------------------------------------------------------------------------}
+
+data Err wid
+    = NoSuchWallet wid
+    | WalletAlreadyExists wid
+    deriving (Show, Eq, Functor, Foldable, Traversable)
+
+errNoSuchWallet :: ErrNoSuchWallet -> Err WalletId
+errNoSuchWallet (ErrNoSuchWallet wid) = NoSuchWallet wid
+
+errWalletAlreadyExists :: ErrWalletAlreadyExists -> Err WalletId
+errWalletAlreadyExists (ErrWalletAlreadyExists wid) = WalletAlreadyExists wid
+
+{-------------------------------------------------------------------------------
+  Mock implementation
+-------------------------------------------------------------------------------}
+
+-- | Shortcut for wallet type.
+type MWallet = Wallet (SeqState DummyTarget) DummyTarget
+
+-- | Mock wallet ID -- simple and easy to read
+newtype MWid = MWid String
+    deriving (Show, Eq, Ord, Generic)
```

Comment:
Same remark here, I don't think the extra indirection is helpful here. It adds another layer of indirection that has  little value? Arbitrary wallet id are easy to generate (especially if just taken from a fixed list of elements) and, this does increase the cognitive load necessary to process this module. Don't you think?